### ISSUE TITLE
Application compile failure on ppc64le with some (but not all) 'vectorize' directives

### ISSUE THREAD

**Issue by cimes-isi** [halide/Halide#5632]
Building on a PowerPC (ppc64le) architecture with LLVM 11.0.0 and Halide master, our application [1] builds a generator but then fails when using it to compile the library with our schedule, which previously worked with Halide v10.0.0.  I could not reproduce this on x86.  Specifically, a generator fails unless the `vectorize` directives at lines 276 and 279 are removed [2].  However, the compilation failure is not actually indicative of what/where the problem is.  Instead it fails somewhere in LLVM:

```
$ make backprojection VERBOSE=1
<snipped>
[ 66%] Generating backprojection.h, backprojection.o
./backprojection.generator -n backprojection -d 0 -g backprojection -f backprojection -e c_header,object -o . target=host-c_plus_plus_name_mangling-large_buffers-no_runtime
Scheduling for CPU: target(powerpc-64-linux-c_plus_plus_name_mangling-large_buffers-no_runtime-power_arch_2_07-vsx)
Block size: 64
Vector size: 16
backprojection.generator: /ccs/home/cimes/casper/llvm-project-11/llvm/lib/IR/Instructions.cpp:453: void llvm::CallInst::init(llvm::FunctionType*, llvm::Value*, llvm::ArrayRef<llvm::Value*>, llvm::ArrayRef<llvm::OperandBundleDefT<llvm::Value*> >, const llvm::Twine&): Assertion `(i >= FTy->getNumParams() || FTy->getParamType(i) == Args[i]->getType()) && "Calling a function with a bad signature!"' failed.
make[3]: *** [backprojection.h] Aborted (core dumped)
make[3]: Leaving directory `/autofs/nccs-svm1_home1/cimes/casper/halide-sar-app/build-master'
```

We originally suspected the problem might be with the name mangling or large buffer features, but removing them did not solve the problem, whereas removing just the two aforementioned `vectorize` directives did, even though other `vectorize` directives remain in the schedule.  Halide should probably print out more detailed information when LLVM fails.

I acknowledge that this is a non-trivial pipeline and perhaps difficult to reproduce.  I will work on creating a simpler example if possible, but I wanted to get the problem out there now in case somebody already has ideas about what might be causing this, which also might help create a narrower test case.

Thanks.

[1] https://github.com/ISI-apex/halide-sar-app
[2] https://github.com/ISI-apex/halide-sar-app/blob/f8f6a945906cc782e9f9c98c600ac64755e8b251/backprojection.cpp#L276-L279

**Comment by abadams** [comment]
My guess is something in #5527 broke PPC. Codegen is trying to call an llvm Function with the wrong number of args. You can see what it's trying to codegen by setting HL_DEBUG_CODEGEN=4. Presumably it's one of the altivec things. That might make it possible to come up with a simple repro.

**Comment by dsharletg** [comment]
CodeGen_PowerPC isn't using the overloaded intrinsic mechanism though, so #5527 shouldn't affect it.

**Comment by dsharletg** [comment]
I can reproduce simd_op_check failures on master. #5531 will affect this a lot, I'm just trying to fix it there. That PR is basically ready to merge, it's just been stumbling over unrelated build bot failures.

**Comment by Infinoid** [comment]
Here is a simple reproducer that runs fine on x86-64, as long as your llvm supports cross-compiling for powerpc.

```c++
#include <stdio.h>
#include <Halide.h>

using namespace Halide;

int main() {
    Var x("x");
    Func f;
    f(x) = max(x, 0);
    // f(x) = max(cast<double>(x), Expr(0.0));
    f.vectorize(x, 16);
    // Target tgt("powerpc-64-linux");
    Target tgt("powerpc-64-linux-vsx");
    f.compile_to_static_library("f", {}, "f", tgt);

    return 0;
}
```

The behavior I'm seeing is:

```
double max, vector size 16, target=powerpc-64-linux     → compiles fine
double max, vector size 16, target=powerpc-64-linux-vsx → function signature error (llvm.ppc.vsx.xvmaxdp)
int    max, vector size 16, target=powerpc-64-linux     → function signature error (llvm.ppc.altivec.vmaxsw)
int    max, vector size 16, target=powerpc-64-linux-vsx → function signature error (llvm.ppc.altivec.vmaxsw)
```

**Comment by dsharletg** [comment]
Thanks for the simplified reproducer. That reproducer works for me on the #5531 branch. I am going to merge it as soon as I can get it to run cleanly on the build bots.

**Comment by Infinoid** [comment]
Looks like @abadams is correct, the issue was introduced in #5527.

I did a `git checkout 42b1a6ed34fbf9586e986afcbc0cd24a78e7e498^` and rebuilt Halide, and was able to build the ISI SAR app successfully for powerpc-64 target.

**Comment by cimes-isi** [comment]
Thanks for everyone's help on this! I can confirm that @Infinoid's reproducer fails for me on both my x86 VM and on the ppc64le hardware.  Changing the Target to "host" on the ppc64le hardware similarly produces the problem.

FYI we had also been able to "work around" the problem by setting the vector size to 1.

I think the credit for identifying the use of `max` as the co-conspirator (along with `vectorize`) here goes to @BachiLi.

**Comment by cimes-isi** [comment]
I can confirm that the problem is resolved (now using revision 0dfdc0d). Thanks again!
